File: 21_07_2024_Build_a_Matrix_With_Conditions.py
class Solution:
    def buildMatrix(self, k: int, rowConditions: list[list[int]], colConditions: list[list[int]]) -> list[list[int]]:
        graph = defaultdict(list)
        for edge in rowConditions:
            graph[edge[0]].append(edge[1])

        row_sorting = self.topo_sort(graph, k)
        graph.clear()
        for edge in colConditions:
            graph[edge[0]].append(edge[1])
        col_sorting = self.topo_sort(graph, k)

        if not row_sorting or not col_sorting:
            return []

        value_position = {}
        for i in range(k):
            value_position[row_sorting[i]] = [i, 0]
            value_position[col_sorting[i]][1] = i

        res = [[0] * k for _ in range(k)]
        for value in range(1, k + 1):
            row = value_position[value][0]
            column = value_position[value][1]
            res[row][column] = value

        return res

    def dfs(self, src: int, graph: defaultdict(list), visited: set, cur_path: set, res: list) -> bool:
        if src in cur_path:
            return False  
        if src in visited:
            return True  

        visited.add(src)
        cur_path.add(src)

        for neighbor in graph[src]:
            if not self.dfs(neighbor, graph, visited, cur_path, res):
                return False

        cur_path.remove(src)
        res.append(src)
        return True

    def topo_sort(self, graph: defaultdict(list), k: int) -> list[int]:
        visited = set()
        cur_path = set()
        res = []

        for src in range(1, k + 1):
            if not self.dfs(src, graph, visited, cur_path, res):
                return []

        res.reverse()
        return res

File: 21_07_2024_Build_a_Matrix_With_Conditions.java
class Solution {
    public int[][] buildMatrix(int k, int[][] rowConditions, int[][] colConditions) {
        Map<Integer, List<Integer>> graph = new HashMap<>();
        for (int[] edge : rowConditions) {
            graph.computeIfAbsent(edge[0], x -> new ArrayList<>()).add(edge[1]);
        }

        List<Integer> rowSorting = topoSort(graph, k);
        graph.clear();
        for (int[] edge : colConditions) {
            graph.computeIfAbsent(edge[0], x -> new ArrayList<>()).add(edge[1]);
        }
        List<Integer> colSorting = topoSort(graph, k);

        if (rowSorting.isEmpty() || colSorting.isEmpty()) {
            return new int[0][0];
        }

        Map<Integer, int[]> valuePosition = new HashMap<>();
        for (int i = 0; i < k; ++i) {
            valuePosition.put(rowSorting.get(i), new int[]{i, -1});
            valuePosition.get(rowSorting.get(i))[1] = i;
        }

        int[][] res = new int[k][k];
        for (int value = 1; value <= k; ++value) {
            int row = valuePosition.get(value)[0];
            int column = valuePosition.get(value)[1];
            res[row][column] = value;
        }

        return res;
    }

    private boolean dfs(int src, Map<Integer, List<Integer>> graph, Set<Integer> visited, Set<Integer> curPath, List<Integer> res) {
        if (curPath.contains(src)) return false;  
        if (visited.contains(src)) return true;  
        
        visited.add(src);
        curPath.add(src);

        for (int neighbor : graph.getOrDefault(src, new ArrayList<>())) {
            if (!dfs(neighbor, graph, visited, curPath, res))
                return false;
        }

        curPath.remove(src);
        res.add(src);
        return true;
    }

    private List<Integer> topoSort(Map<Integer, List<Integer>> graph, int k) {
        Set<Integer> visited = new HashSet<>();
        Set<Integer> curPath = new HashSet<>();
        List<Integer> res = new ArrayList<>();

        for (int src = 1; src <= k; ++src) {
            if (!dfs(src, graph, visited, curPath, res))
                return Collections.emptyList();
        }

        Collections.reverse(res);
        return res;
    }
}

